torch/csrc/jit/passes/constant_propagation.cpp

```diff
@@ -63,28 +61,124 @@ void propagateNode(Node* n) {
+  };
+
+  for (Node *orig : body->nodes()) {
+    Node *clone = graph->insertNode(graph->createClone(orig, get_value));
+    for (size_t i = 0; i < orig->outputs().size(); ++i) {
+      value_map[orig->outputs()[i]] = clone->outputs()[i];
+    }
+  }
+  for (size_t i = 0; i < n->outputs().size(); ++i) {
+    n->outputs().at(i)->replaceAllUsesWith(get_value(body->outputs().at(i)));
+  }
+  // NB: destroy the node here, because it might contain side effects, like print
+  n->destroy();
+}
+
+bool isTrueConstant(Value *val) {
+  at::optional<bool> maybe_value = constant_as<bool>(val);
+  return maybe_value && *maybe_value;
+}
+
+void lowerIf(Node *n) {
+  if (isTrueConstant(n->input())) {
+    lowerIf(n->blocks()[0], n);
+  } else {
+    lowerIf(n->blocks()[1], n);
+  }
+}
+
+//returns true if the mutated variables are changed
+bool recomputeMutatedVariables(Node *n) {
+  JIT_ASSERTM(n->kind() == prim::If, "Only supported for If nodes");
+  std::unordered_set<Value*> mutated_variables;
+  for (Block * block : n->blocks()) {
+    for (Node *n : block->nodes()) {
+      for (size_t i = 0; i < n->outputs().size(); ++i) {
+        mutated_variables.insert(n->outputs()[i]);
+      }
+    }
+  }
+  auto true_block = n->blocks()[0];
+  auto false_block = n->blocks()[1];
+  auto initial_outputs = true_block->outputs().size();
+  for (size_t i = 0; i < true_block->outputs().size();) {
+    //neither block mutates output i
+    if (!mutated_variables.count(true_block->outputs()[i]) &&
+      !mutated_variables.count(false_block->outputs()[i])) {
+      n->outputs().at(i)->replaceAllUsesWith(true_block->outputs()[i]);
+      n->eraseOutput(i);
+      true_block->eraseOutput(i);
+      false_block->eraseOutput(i);
+    } else {
+      i++; //increment bc we didn't remove current index
+    }
+  }
+  //an output was removed
+  return initial_outputs != true_block->outputs().size();
+}
+
 } // anonymous namespace
 
-void ConstantPropagation(Node* n, bool recurse) {
+//returns whether the node's set of mutated variables changed
+bool ConstantPropagation(Node* n, bool recurse) {
   bool constant_inputs = (n->inputs().size() > 0) &&
     std::all_of(n->inputs().begin(), n->inputs().end(), [&](Value* v) {
       return v->node()->kind() == prim::Constant;
     });
   bool supported_node = skip_list.count(n->kind()) == 0;
-  if (constant_inputs && supported_node) {
+  auto run_blocks = [&]() {
+    bool any_child = false;
+    if (recurse) {
+      for (Block * block : n->blocks()) {
+        auto child = ConstantPropagation(block, recurse);
+        any_child = any_child || child;
+      }
+    }
+    return any_child;
+  };
+  if (n->kind() == prim::If) {
+    //did a child node change
+    bool changed = run_blocks();
+    //inline node if we can, otherwise if a child node changed recompute
+    //mutated variables and see if this node changed
+    if (constant_inputs) {
+      lowerIf(n);
+    } else if (changed) {
+      changed = recomputeMutatedVariables(n);
```

Comment:
If it is possible to simplify recomputeMutatedVariables so it just has to check the output list of the if, then it is also ok to remove the `changed` logic and just unconditionally check for simplified outputs.